storage/remote/client.go

```diff
@@ -257,22 +262,36 @@ func (c *client) Read(ctx context.Context, query *prompb.Query) (*prompb.QueryRe
 	if err != nil {
 		return nil, errors.Wrap(err, "error sending request")
 	}
+
+	contentType := httpResp.Header.Get("Content-Type")
+	if strings.HasPrefix(contentType, "application/x-protobuf") {
+		return c.handleSampledResponse(httpResp, start, req, sortSeries)
+	}
+
+	if httpResp.StatusCode/100 != 2 {
+		io.Copy(ioutil.Discard, httpResp.Body)
+		httpResp.Body.Close()
```

Comment:
I hope this is what you meant.

Reply:
Why not having:

```
	defer func() {
		io.Copy(ioutil.Discard, httpResp.Body)
		httpResp.Body.Close()
}()
```
here?

Reply:
Really sorry Bartek. Missed the notification.

Reply:
Still not addressed (:

Reply:
still not addressed, in case below with unsupported contentType the response body will be not closed/exhausted leaving the connection alive without good reason.

Reply:
Sorry I got confused with this and closing the HTTP connection for the streamed use-case. 
I solved the issue of closing the `streaming` HTTP connection and considered this ^ issue solved too.
```
res := &prompb.ChunkedReadResponse{}
	if err := c.stream.NextProto(res); err == io.EOF {
		if c.httpResp != nil {
			c.httpResp.Body.Close()
		}
		return false
```

Reply:
To solve the problem with the unsupported contentType, is the additional line here enough to solve the issue?
```
	if !strings.HasPrefix(contentType, "application/x-streamed-protobuf; proto=prometheus.ChunkedReadResponse") {
-->		httpResp.Body.Close()
		return nil, errors.Errorf("not supported remote read content type: %s", contentType)
	}
```

Reply:
Hmm, I considered the above option but won't it be incorrect? 
for eg: We're going to return `return &chunkedResponseSeriesSet{stream: stream, httpResp: httpResp}, nil` in `StartRead()` and thus if we have 
```	
defer func() {
		io.Copy(ioutil.Discard, httpResp.Body)
		httpResp.Body.Close()
        }()
```
before the `return &chunkedResponseSeriesSet{stream: stream, httpResp: httpResp}, nil` we will end up closing the `http` connection that is needed for `stream.NextProto(res)`. 
From my understanding, the most straightforward option seems to be to close the `HttpRest` individually/separately depending on the use case.

Reply:
well there is better way. You are right that we cannot do it for the case when we return iterator. So in this case I would do this.:

```
defer func() {
            if err != nil {
		io.Copy(ioutil.Discard, httpResp.Body)
		httpResp.Body.Close()
              }
        }()
```
Then make sure the function returns `(<something>, err error)` and it will only close on error (:

Reply:
This makes sense to me. (I have a small clarifying question though)
But, I can see that won't be returning an error when we see an `io.EOF` in the `chunkedResponseSeriesSet` `Next()`. The connection needs to be closed in this case too, right?
Should the `Next()` be replaced from 
```
if err := c.stream.NextProto(res); err != nil {
		if err != io.EOF {
			c.err = err
		}
		return false
	}
```
Replaced to 
```
if err := c.stream.NextProto(res); err != nil {
		c.err = err
		return false
	}
```
This above approach produces some errors like:
`api_test.go:1784: Unexpected error: execution: EOF`
Or is this the best version:
```
if err := c.stream.NextProto(res); err != nil {
                httpResp.Body.Close()		
                if err != io.EOF {
			c.err = err
		}
		return false
	}
```